### ISSUE TITLE
Provide efficient implementation for operations on lists of tensors.

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#38655]
We should have efficient implementations for a small subset of operations on the lists of tensors, such as 
```
_tensor_list_add.Tensor(Tensor[] self, Tensor[] other, *, Scalar alpha=1)
```
## Motivation

For a lot of GPU training workloads, the performance of the optimizer becomes the bottleneck. Currently there are two extreme cases of gpu optimizer performance - very flexible and hackable implementations in torch.optim, and fully fused rigid implementations in apex (github.com/nvidia/apex), available for sgd, Adam, and, as of couple days back, adagrad. People 
who want to experiment with optimizers either have to suffer through bad performance that typically would be seen when implementing optimizer the usual way, iterating over lists of tensors, or, if they are up to it, locally modify apex to have a fast fused kernel. The latter part can be pretty challenging. 

A typical optimizer launches 5-10 kernels per parameter tensor. Models can have hundreds of parameters, so 500-1000 kernels have to be launched. Those kernels are often very small, and the operations become cpu-bound. If we can launch a single kernel that operates on all the parameter tensors at the same time, we can reduce the number of kernels launches by a factor of 100 and relieve the cpu-boundness. 

Other situations where efficient implementation of operations on tensor lists are helpful are `.zero_grad()` function, unscaling gradients and checking for `nan` values in the mixed precision training, computing norm of parameters for `clip_grad_norm`, converting parameters/gradients between low- and high- precision type for mixed precision training. 

Nvidia has kernels that can efficiently operate on disjoint list of tensors in apex, but it's impossible to use those kernels outside of apex, and each particular operation has to be implemented (currently those kernels are used for mixed precision gradient scaling and a few fused optimizers). If efficient list operations were available in pytorch core, that would
enable much easier experimentation. 

## Proposal

Expose experimental tensor list operations for arithmetic operations, `fill_`, `norm` backed by efficient implementations.
Initial list of operations can be discussed, but tentatively, it should be enough to implement operations in the existing optimizers. In particular, we should support inplace variants, and operations between tensor list and scalar. E.g., for `add` we want the 
following variants:
```
_add_tensor_list.Tensor(Tensor[] self, Tensor other[], *, Scalar alpha=1) -> Tensor[]
_add_tensor_list.Scalar(Tensor[] self, Scalar other, Scalar alpha=1) -> Tensor[]
_add_tensor_list_.Tensor(Tensor[](a!) self, Tensor[] other, *, Scalar alpha=1) -> Tensor[](a!) #inplace
_add_tensor_list_.Scalar(Tensor[](a!) self, Scalar other, Scalar alpha=1) -> Tensor[](a!) #inplace scalar variant
_add_tensor_list.out(Tensor[] self, Tensor[] other, *, Scalar alpha=1, Tensor[](a!) out) -> Tensor[](a!)
```
Type promotion should be supported as for regular operations. 
Shapes in the list arguments should be the same, error out on mismatch or if broadcasting is required (this is open for debate).
For backends that don't have efficient implementations or don't need them fall back to implemetations looping over tensors in the list. 

After this is done, existing optimizers can be rewritten to use these ops, leading to better performance. Optimizer implementation itself will still remain readable, accessible and hackable.

Backward formulas can be provided for operations on the list. However, given that primary focus is use in optimizers, backward should not be a blocker. 

## Alternatives

This proposal has a significant overlap with NestedTensor proposal, but we understand that it is hard to implement NestedTensors to cover all the relevant scenarios. This proposal is much more limited in scope, and efficient implementations that are put in the core as part of this proposal can be reused for NestedTensor. 
Similarly, once efficient implementations are in the core, TorchScript can have passes that will convert operations that loop over tensors in the list to call into these efficient implementations. 

cc @cpuhrsch @vincentqb @mcarilli @soumith @dzhulgakov @ezyang

**Comment by mcarilli** [comment]
Awesome, been looking forward to getting this in core for a long time.  Do we want the implementations to handle arbitrary tensor layouts, or layouts that don't match across across or between lists?  I think it must, for consistency with TensorIterator-based APIs, which do allow ie a+b where a and b have different layouts.

**Comment by ngimel** [comment]
For consistency it should handle non-matching layouts, but it's ok to call .contiguous() on problematic tensors rather than torture the kernels themselves. Matching memory-dense (but not necessarily contiguous) tensors in the lists should not be a problem, right?

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/27522

**Comment by soumith** [comment]
`_add_tensor_list` as a private function makes sense, and from what I can see the work on the kernels can be reused to any final / better implementation.

Let's do this!

**Comment by mcarilli** [comment]
For reference my current harness is here: https://github.com/NVIDIA/apex/blob/master/csrc/multi_tensor_apply.cuh.  It accepts a functor that allows arbitrary pointwise ops on lists of tensors, and processes them in one or a few launches with bandwidth comparable to one pointwise launch on a single flattened tensor.  It can also be used for reductions.  Several people (@slayton58, @FDecaYed, @thorjohnsen, @tulloch, @crcrpar) have used it to create custom fused optimizers.

**Comment by dzhulgakov** [comment]
Yay, let's do it!

How would the re-written optimizers look like? Explicit calls to `torch._tensor_list_add` and others  might be a bit of a mouthful. And if we wrap it into some python wrapper with operator overloading it might not work with TorchScript.

**Comment by mcarilli** [comment]
@dzhulgakov On the optimizer side, the only minor annoyance is organizing params+grads+state buffers into lists according to how flexible the multi tensor apply backend is.  For example, the Apex implementation only handles lists as long as all tensors in each list are the same dtype**, so params+grads need to be split accordingly ([see @ajtulloch 's code](https://github.com/NVIDIA/apex/blob/3bae8c83494184673f01f3867fa051518e930895/apex/optimizers/fused_adagrad.py#L80-L101)).  I think it remains legible.  The split across dtypes could even be moved to the backend so the user can pass any tensor list.

** for ops that act on several lists, the dtype must be the same within each list but may differ across lists.  This enables me to effectively fuse casting with any pointwise op, ie, read list of fp16 tensors, perform arbitrary pointwise op, and write out into a list of fp32 tensors.  Also, fun fact, if the output tensor list was created as sliced views of a large empty tensor, what you've done is a fused cast + pointwise op + cat.  Multi tensor apply is a great way to do cat, cleaner than the existing implementation in that no auxiliary metadata tensors need to be allocated.  Metadata is passed via kernel args.

**Comment by gchanan** [comment]
> Awesome, been looking forward to getting this in core for a long time. Do we want the implementations to handle arbitrary tensor layouts, or layouts that don't match across across or between lists? I think it must, for consistency with TensorIterator-based APIs, which do allow ie a+b where a and b have different layouts.

what definition of layout are we using here?  strides?

> Shapes in the list arguments should be the same, error out on mismatch or if broadcasting is required (this is open for debate).

this is a bit of bike-shedding, but it's not obvious what the semantics are when we call things like `_add_tensor_list` -- it sounds like tensor_list is some opaque thing with its own semantics (like NestedTensor) when I don't think we want to start getting in that general game.

AFAICT, up until @mcarilli's point about doing "fused cast + pointwise op + cat" we are really just talking about doing a "for each" function application -- and even @mcarilli's optimization would maybe fit (although I'm not sure whether autograd would actually work given the limitations on views).

So, long story short, it seems nicer to call things "_add_foreach.Tensor[]" or "_add_foreach.TensorList" or similar.

**Comment by ngimel** [comment]
Agreed, _add_foreach is a better name, we definitely don't want opaque things. 
Layout here means strides, yes.

**Comment by cpuhrsch** [comment]
This is great! NestedTensor can lower a lot of operations to TensorList. In fact, the TensorList operations NestedTensor stand to launch will be a subset of the ones planned to be supported here. NestedTensor is a Tensor-esque type and as such must have a consistent dtype, device, dimension and layout. So, it's not the right container for this case, if we want to support mixed dtypes, layouts etc.

One comment here, you might want the Tensors in a given TensorList to be part of a large contiguous piece of memory. This is a condition that could be detected at runtime based on data_ptr, offsets, numel() etc., but there could also be place for a PackedTensorList structure or such that has some additional properties along these lines. This is again something NestedTensor could dispatch to.

**Comment by FDecaYed** [comment]
This is great news!
One thing to note is launching single kernel over a list of tensor is not only a CPU optimization, but also help achieving higher bandwidth on device side. 
So for workloads thats are not CPU bound, probably thanks to the async nature of work scheduling, this can still benefit them.

**Comment by mcarilli** [comment]
@cpuhrsch I don't think Greg means "TensorList" will be its own object.  This is not something intended to rival NestedTensor, it's lower-level and can serve as an implementation detail of NestedTensor someday.  I think Greg means `_add_foreach.TensorList` is what the overload declaration will look like in native_functions.yaml and `TensorList` is `ArrayRef<Tensor>`.

**Comment by vadimkantorov** [comment]
> One comment here, you might want the Tensors in a given TensorList to be part of a large contiguous piece of memory.

If `flatten_parameters` was supported on general models, that could be the result of it :)

**Comment by ngimel** [comment]
With current multi_tensor_apply implementation in apex there's very little difference between operating on a large contiguous space of memory or on the tensors in the different memory locations. It is still good to keep in mind as potential optimization, but the general idea is to not require `flatten_parameters` or something similar (which is memory and performance overhead) and still be able to operate on parameters/gradients in an efficient way.

**Comment by cpuhrsch** [comment]
> @cpuhrsch I don't think Greg means "TensorList" will be its own class, if that's what you meant. This is not something intended to rival NestedTensor, it's lower-level and can serve as an implementation detail of NestedTensor someday. I think Greg means `_add_foreach.TensorList` is what the overload declaration will look like in native_functions.yaml and `TensorList` is `ArrayRef<Tensor>`.

**Comment by SsnL** [comment]
@ngimel I'm curious about the performance tradeoff between this and performing single kernel ops on a giant flat vector containing all parameters (i.e., the actual parameters would be non-overlapping views of it).

**Comment by ngimel** [comment]
The benchmarks show that multi-tensor-apply kernels achieve high fraction of peak bandwidth (@mcarilli will know more). The problem with giant flat vector containing params is that it's usually pretty fragile (unless you have a simple training loop, always use all the parameters, no parameter sharing etc).

**Comment by vadimkantorov** [comment]
One way to support flatten_parameters would be checking if all inputs to multi_tensor_apply are indeed contiguous in memory (a complication may be if there are any memory alignment gutters), and then launching a single kernel. Then an optional flatten_parameters may be supported to be explicitly used only for advanced cases.

**Comment by SsnL** [comment]
@ngimel Nice, thanks for responding quickly. Since the conditions for giant flat vector is actually met for most use cases, it is great to hear that we will have a solution that doesn't sacrifice much performance for generality.

**Comment by mcarilli** [comment]
> I'm curious about the performance tradeoff between this and performing single kernel ops on a giant flat vector containing all parameters (i.e., the actual parameters would be non-overlapping views of it).

MTA kernels should achieve bandwidth comparable to acting on a giant flat vector.  If you find a realistic case where they don't, let us know.

**Comment by SsnL** [comment]
@mcarilli That's great to hear. Will do.

**Comment by mcarilli** [comment]
Optimizers should robustly handle the cases where not all parameters receive grads every iteration.
Updated with a simple filter-helper example I proposed to our mlperf people to reduce MTA list-packing overhead for their fused optimizers.
I'll probably PR a C++ list-packing helper like that for MTA amp.GradScaler.unscale_ soon.

**Comment by vadimkantorov** [comment]
There's also now some FlatParameter in FSDP. Is it related to general parameter flattening?

**Comment by awgu** [comment]
> There's also now some FlatParameter in FSDP. Is it related to general parameter flattening?

I do not think that `FlatParameter` in FSDP is related to general parameter flattening. It is actually just a normal `nn.Parameter` (that does not even override a single method).
https://github.com/pytorch/pytorch/blob/14d5f139d205f924eb7ddd3e61215971bd194855/torch/distributed/fsdp/flat_param.py#L123

We have specialized `FlatParameter` to represent something in the context of FSDP (adding attributes to it and using it in a specific way). However, with respect to `torch` functions, it is just a 1D `nn.Parameter`.
